File: 13.DFS.BFS/22.两段击/22.括号生成.java
/*
 * @lc app=leetcode.cn id=22 lang=java
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    class Node {
        public String path;
        public int lcnt = 0;
        public int rcnt = 0;
        public Node() {}
    }
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        if (n <= 0) {
            return ans;
        }

        List<Node> cur = new ArrayList<>();
        Node x = new Node();
        x.path = "(";
        x.lcnt = 1;
        x.rcnt = 0;
        cur.add(x);

        while (cur.size() > 0) {
            List<Node> tmp = new ArrayList<>();
            for (Node c: cur) {
                if (c.path.length() == (n+n) && c.lcnt == c.rcnt) {
                    ans.add(c.path);
                    continue;
                }

                if (c.lcnt >= c.rcnt && c.lcnt < n) {
                    Node next = new Node();
                    next.path = c.path + "(";
                    next.lcnt = c.lcnt + 1;
                    next.rcnt = c.rcnt;
                    tmp.add(next);
                }

                if (c.lcnt > c.rcnt) {
                    Node next = new Node();
                    next.path = c.path + ")";
                    next.lcnt = c.lcnt;
                    next.rcnt = c.rcnt + 1;
                    tmp.add(next);
                }
            }

            cur = tmp;
        }
        return ans;
    }
}
// @lc code=end

File: 13.DFS.BFS/22.两段击/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    struct node {
        string path;
        int lcnt = 0;
        int rcnt = 0;
    };

   public:
    vector<string> generateParenthesis(int n) {
        if (n <= 0) {
            return {};
        }

        vector<string> ans;

        vector<node> cur, tmp;

        node t;
        t.path = "(";
        t.lcnt = 1;

        cur.push_back(t);

        for (int i = 1; i < (n + n); i++) {
            tmp.clear();

            for (auto &c : cur) {
                if (c.lcnt >= c.rcnt && c.lcnt < n) {
                    node next = c;
                    next.path.push_back('(');
                    next.lcnt++;
                    tmp.push_back(next);
                }

                if (c.lcnt > c.rcnt) {
                    node next = c;
                    next.path.push_back(')');
                    next.rcnt++;
                    tmp.push_back(next);
                }
            }

            cur.swap(tmp);
        }

        for (auto &c : cur) {
            ans.push_back(c.path);
        }

        return ans;
    }
};
// @lc code=end
